File: apps/web/src/lib/genres.ts
// Genre catalogue for the interests / hidden-genres picker. The `name` doubles as
// the value sent to /api/v1/podcasts/discover?category=… and matched against a
// podcast's own categories for filtering. This mirrors Apple Podcasts' top-level
// categories (which Podcast Index also understands).
//
// `name` is therefore a protocol value and must stay English. `id` is the
// translation key used for display — see `genreLabel`.

import { t, type MessageKey } from '$lib/i18n';

export interface Genre {
	/** Translation key suffix under `genres.*`. Display only. */
	id: string;
	/** Wire value sent to the API. Never translate. */
	name: string;
	icon: string;
}

export const GENRES: Genre[] = [
	{ id: 'arts', name: 'Arts', icon: 'ph-palette' },
	{ id: 'business', name: 'Business', icon: 'ph-briefcase' },
	{ id: 'comedy', name: 'Comedy', icon: 'ph-smiley' },
	{ id: 'education', name: 'Education', icon: 'ph-graduation-cap' },
	{ id: 'fiction', name: 'Fiction', icon: 'ph-book-open' },
	{ id: 'government', name: 'Government', icon: 'ph-bank' },
	{ id: 'healthFitness', name: 'Health & Fitness', icon: 'ph-heartbeat' },
	{ id: 'history', name: 'History', icon: 'ph-scroll' },
	{ id: 'kidsFamily', name: 'Kids & Family', icon: 'ph-baby' },
	{ id: 'leisure', name: 'Leisure', icon: 'ph-game-controller' },
	{ id: 'music', name: 'Music', icon: 'ph-music-notes' },
	{ id: 'news', name: 'News', icon: 'ph-newspaper' },
	{ id: 'religionSpirituality', name: 'Religion & Spirituality', icon: 'ph-hands-praying' },
	{ id: 'science', name: 'Science', icon: 'ph-atom' },
	{ id: 'societyCulture', name: 'Society & Culture', icon: 'ph-users-three' },
	{ id: 'sports', name: 'Sports', icon: 'ph-basketball' },
	{ id: 'technology', name: 'Technology', icon: 'ph-cpu' },
	{ id: 'tvFilm', name: 'TV & Film', icon: 'ph-film-slate' },
	{ id: 'trueCrime', name: 'True Crime', icon: 'ph-fingerprint' }
];

const BY_NAME = new Map(GENRES.map((g) => [g.name.toLowerCase(), g]));

/**
 * Display label for a genre. Takes the wire value ("True Crime") and returns the
 * translated name. Categories that came from a feed and are not in our catalogue
 * are passed through untouched — a publisher's own category is not ours to
 * rewrite.
 */
export function genreLabel(name: string): string {
	if (!name) return '';
	if (name.toLowerCase() === 'all') return t('genres.all');
	const genre = BY_NAME.get(name.toLowerCase());
	return genre ? t(`genres.${genre.id}` as MessageKey) : name;
}

File: apps/android/core/ui/src/main/kotlin/net/koalastuff/koalacast/core/ui/genre/Genres.kt
package net.koalastuff.koalacast.core.ui.genre

import androidx.annotation.StringRes
import net.koalastuff.koalacast.core.ui.R

/**
 * Mirrors `apps/web/src/lib/genres.ts` — Apple Podcasts' top-level categories, which
 * Podcast Index also understands.
 *
 * [wireName] is a protocol value sent as `?category=` and matched against a feed's own
 * categories. It must stay English and must never be translated; [labelRes] is the
 * display side. Keep this list in step with the web catalogue.
 */
data class Genre(
    val id: String,
    val wireName: String,
    @StringRes val labelRes: Int,
)

val GENRES: List<Genre> = listOf(
    Genre("arts", "Arts", R.string.genre_arts),
    Genre("business", "Business", R.string.genre_business),
    Genre("comedy", "Comedy", R.string.genre_comedy),
    Genre("education", "Education", R.string.genre_education),
    Genre("fiction", "Fiction", R.string.genre_fiction),
    Genre("government", "Government", R.string.genre_government),
    Genre("healthFitness", "Health & Fitness", R.string.genre_health_fitness),
    Genre("history", "History", R.string.genre_history),
    Genre("kidsFamily", "Kids & Family", R.string.genre_kids_family),
    Genre("leisure", "Leisure", R.string.genre_leisure),
    Genre("music", "Music", R.string.genre_music),
    Genre("news", "News", R.string.genre_news),
    Genre("religionSpirituality", "Religion & Spirituality", R.string.genre_religion_spirituality),
    Genre("science", "Science", R.string.genre_science),
    Genre("societyCulture", "Society & Culture", R.string.genre_society_culture),
    Genre("sports", "Sports", R.string.genre_sports),
    Genre("technology", "Technology", R.string.genre_technology),
    Genre("tvFilm", "TV & Film", R.string.genre_tv_film),
    Genre("trueCrime", "True Crime", R.string.genre_true_crime),
)

private val BY_WIRE_NAME = GENRES.associateBy { it.wireName.lowercase() }

/**
 * The label resource for a wire value, or `null` when the feed supplied a category of
 * its own — a publisher's category is not ours to rewrite, so it is shown verbatim.
 */
@StringRes
fun genreLabelRes(wireName: String): Int? = BY_WIRE_NAME[wireName.lowercase()]?.labelRes
